src/Workspaces/Core/Portable/EmbeddedLanguages/VirtualChars/VirtualChar.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars
+{
+    /// <summary>
+    /// The Regex and Json parsers wants to work over an array of characters, however this array of
+    /// characters is not the same as the array of characters a user types into a string in C# or
+    /// VB. For example In C# someone may write: @"\z".  This should appear to the user the same as
+    /// if they wrote "\\z" and the same as "\\\u007a".  However, as these all have wildly different
+    /// presentations for the user, there needs to be a way to map back the characters it sees ( '\'
+    /// and 'z' ) back to the  ranges of characters the user wrote.  
+    ///
+    /// VirtualChar serves this purpose.  It contains the interpreted value of any language
+    /// character/character-escape-sequence, as well as the original SourceText span where that
+    /// interpreted character was created from.  This allows the regex and json parsers to both
+    /// process input from any language uniformly, but then also produce trees and diagnostics that
+    /// map back properly to the original source text locations that make sense to the user.
+    /// </summary>
+    internal readonly struct VirtualChar : IEquatable<VirtualChar>
+    {
+        public readonly char Char;
```

Comment:
Seems like rather inefficient representation for long strings that have just a couple of escape characters.

Reply:
I also wonder how is this gonna work with UTF8 string feature.

Reply:
@jaredpar Is UTF8 string literal proposal being worked on?